### PR TITLE
Bug 1868539 - Add method of changing remote server url

### PR DESCRIPTION
**Pull request by sarah541**
| Fenix  | Focus |
| ------------- | ------------- |
|   |   |

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1868539

### CODE REVIEW

File: fenix/app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -735,6 +735,12 @@
     <!-- Instructions on how to access pairing -->
     <string name="pair_instructions_2"><![CDATA[Scan the QR code shown at <b>firefox.com/pair</b>]]></string>
 
+    <!-- Serp telemetry url Preferences -->
```
Comment (gabrielluong) [inline fenix/app/src/main/res/values/strings.xml:738]:
```suggestion
    <!-- SERP telemetry url Preferences -->
```
Let's just capitalize SERP everywhere. This comment applies. throughout the whole PR changes so it might be worth looking back at your changes here.

File: fenix/app/src/main/java/org/mozilla/fenix/components/Core.kt [hunk 2]
```diff
@@ -322,6 +322,11 @@ class Core(
                         rootStorageDirectory = context.filesDir,
                         readJson = readJson,
                         collectionName = COLLECTION_NAME,
+                        serverUrl = if (context.settings().serpTelemetryServer) {
+                            REMOTE_MAIN_ENDPOINT_URL
+                        } else {
+                            REMOTE_TEST_ENDPOINT_URL
+                        },
```
Comment (MatthewTighe) [inline fenix/app/src/main/java/org/mozilla/fenix/components/Core.kt:329]: If we turned this into a lambda, we could change it during runtime. We would also need to rebuild the `RemoteSettingsClient`, but I think that should be fine. For example

```
class SerpTelemetryRepository(
    rootStorageDirectory: File,
    private val readJson: () -> JSONObject,
    collectionName: String,
    serverUrl: { String } = { REMOTE_MAIN_ENDPOINT_URL, }
)

// here
serverUrl = { if (context.settings().serpTelemetryServer) {
        REMOTE_MAIN_ENDPOINT_URL
    } else {
        REMOTE_TEST_ENDPOINT_URL
    },
}

File: focus-android/app/src/main/java/org/mozilla/focus/settings/advanced/SecretSettingsFragment.kt [hunk 3]
```diff
@@ -39,6 +43,24 @@ class SecretSettingsFragment :
         }
     }
 
+    private fun setupSERPTelemetryServer() {
```
Comment (MatthewTighe) [inline focus-android/app/src/main/java/org/mozilla/focus/settings/advanced/SecretSettingsFragment.kt:46]: Contrary to GL's point, let's stay consistent with acronym cases in code. Would still be good to capitalize it in comments and strings though
```suggestion
    private fun setupSerpTelemetryServer() {
```

**Review by MatthewTighe — COMMENTED** [review]
Just a quick pass today. My main question is: have you considered adding this to the debug drawer instead of secret settings? That along with changes so that we can update the server at runtime seem like it would be the easiest usecase for manual testing

File: fenix/app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -735,6 +735,12 @@
...
+    <!-- Preference for using main serp telemetry url -->
+    <string name="preference_main_server">Main</string>
+    <!-- Preference for using test serp telemetry url -->
+    <string name="preference_test_server">Test</string>
```
Comment (linabutler) [inline fenix/app/src/main/res/values/strings.xml:742]:
```suggestion
    <string name="preference_remote_settings_prod_server">Prod</string>
    <!-- Preference for using test serp telemetry url -->
    <string name="preference_remote_settings_stage_server">Stage</string>
```

Ditto here, let's use the "Prod" and "Stage" naming. This would also be consistent with the [Remote Settings DevTools add-on](https://github.com/mozilla-extensions/remote-settings-devtools)!

Screenshot 2024-01-29 at 08 00 36

**Review by linabutler — CHANGES_REQUESTED** [review]

**Comment by sarah541** [comment]
> Just a quick pass today. My main question is: have you considered adding this to the debug drawer instead of secret settings? That along with changes so that we can update the server at runtime seem like it would be the easiest usecase for manual testing

I haven't! That's a good redirection. I can try making use of that.

File: fenix/app/src/main/res/values/static_strings.xml [hunk 5]
```diff
@@ -34,6 +34,8 @@
     <string name="preferences_debug_settings">Secret Settings</string>
     <!-- Label for the secret settings preference -->
     <string name="preferences_debug_info" translatable="false">Secret Debug Info</string>
+    <!-- Label for using serp telemetry server url -->
+    <string name="preferences_debug_settings_serp_telemetry_server_url" translatable="false">SERP telemetry server url (requires restart)</string>
```
[the remark this answers is not in the document]
Comment (sarah541) [reply fenix/app/src/main/res/values/static_strings.xml:38]: Totally works for us if we are using the same urls xD

File: android-components/components/feature/search/src/main/java/mozilla/components/feature/search/telemetry/SerpTelemetryRepository.kt [hunk 6]
```diff
@@ -26,7 +26,7 @@ class SerpTelemetryRepository(
     rootStorageDirectory: File,
     private val readJson: () -> JSONObject,
     collectionName: String,
-    serverUrl: String = REMOTE_ENDPOINT_URL,
+    serverUrl: () -> String = { REMOTE_PROD_ENDPOINT_URL },
```
Comment (MatthewTighe) [inline android-components/components/feature/search/src/main/java/mozilla/components/feature/search/telemetry/SerpTelemetryRepository.kt:29]: Since this parameter is a lambda, can we update the name to communicate that it's doing something? `getServerUrl` for example.

Rather than pass the `serverUrl` to the `RemoteSettingsClient` in this class, we should rebuild the client every time we go to use it so that it can change during runtime. I would suggest a small factory function for this. Then, your use cases would look something like:

```
fun buildClient() = RemoteSettingsClient(getServerUrl(), bucketName, collectionName, storageRootDirectory)

internal suspend fun writeToCache(records: RemoteSettingsResponse): RemoteSettingsResult {
    return buildClient().write(records)
}
```

**Review by MatthewTighe — COMMENTED** [review]
The code around remote settings looks good to me, just one change to request and then I'll return for a review. Tagging @MozillaNoah to take a look at the secret settings piece since he has more recent context. Sarah and I did discuss moving this to debug drawer but decided to migrate it there as a follow-up

**Comment by MatthewTighe** [comment]
Sarah and I chatted offline and she pointed out that since we are only constructing the `BaseSearchTelemetry` class hierarchy at app startup that we currently can't have dynamic updates to the server anyway, so the previous comments about it are no longer valid.

File: android-components/components/feature/search/src/main/java/mozilla/components/feature/search/telemetry/SerpTelemetryRepository.kt (hunk 6 above)
Comment (sarah541) [reply android-components/components/feature/search/src/main/java/mozilla/components/feature/search/telemetry/SerpTelemetryRepository.kt:29]: resolving based on Matt's below comment

File: android-components/components/support/remotesettings/src/main/java/mozilla/components/support/remotesettings/RemoteSettingsClient.kt [hunk 7]
```diff
@@ -43,17 +43,17 @@ import java.net.URL
 
 class RemoteSettingsClient(
     private val storageRootDirectory: File,
-    private val serverUrl: String = "https://firefox.settings.services.mozilla.com",
+    private val serverUrl: () -> String = { "https://firefox.settings.services.mozilla.com" },
```
[the remark this answers is not in the document]
Comment (sarah541) [reply android-components/components/support/remotesettings/src/main/java/mozilla/components/support/remotesettings/RemoteSettingsClient.kt:46]: resolving based on Matt's below comment

File: fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt [hunk 8]
```diff
@@ -129,6 +130,28 @@ class SecretSettingsFragment : PreferenceFragmentCompat() {
...
+            isVisible = Config.channel.isNightlyOrDebug
+        }
+        stagePreference.onClickListener {
+        }
+
+        val serverUrlPreference = requireContext().settings().remoteSettingsServer
+        prodPreference.setCheckedWithoutClickListener(serverUrlPreference)
+        stagePreference.setCheckedWithoutClickListener(!serverUrlPreference)
+
+        addToRadioGroup(prodPreference, stagePreference)
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:154]: Since the preference is a boolean and there's just a binary choice between two radio buttons, would making this a switch preference simplify all of this? 

"Use Production server (staging will be used when disabled)"
- On - production
- Off - staging

File: fenix/app/src/main/java/org/mozilla/fenix/components/Core.kt [hunk 9]
```diff
@@ -623,5 +630,7 @@ class Core(
 
         // collection name to fetch from server for SERP telemetry
         const val COLLECTION_NAME = "search-telemetry-v2"
+        internal const val REMOTE_PROD_ENDPOINT_URL = "https://firefox.settings.services.mozilla.com"
+        internal const val REMOTE_STAGE_ENDPOINT_URL = "https://firefox.settings.services.allizom.org"
```
[the remark this answers is not in the document]
Comment (sarah541) [reply fenix/app/src/main/java/org/mozilla/fenix/components/Core.kt:632]: There is one in AC which serves a fallback purpose, one at the client side as needed.

File: fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt (hunk 8 above)
Comment (sarah541) [reply fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:154]: Yes that makes sense.

File: android-components/components/lib/crash/docs/metrics.md [hunk 10]
```diff
@@ -62,5 +62,5 @@ In addition to those built-in metrics, the following metrics are added to the pi
 
 Data categories are [defined here](https://wiki.mozilla.org/Firefox/Data_Collection).
 
-<!-- AUTOGENERATED BY glean_parser v10.0.3. DO NOT EDIT. -->
```
Comment (MozillaNoah) [inline android-components/components/lib/crash/docs/metrics.md:65]: accidental commit?

File: android-components/components/support/remotesettings/src/main/java/mozilla/components/support/remotesettings/RemoteSettingsClient.kt [hunk 11]
```diff
@@ -63,6 +63,7 @@ class RemoteSettingsClient(
      * Fetches a response that includes Remote Settings records and the last time the collection was modified.
      */
     suspend fun fetch(): RemoteSettingsResult = withContext(Dispatchers.IO) {
+        println("Sarah server url $serverUrl")
```
Comment (MozillaNoah) [inline android-components/components/support/remotesettings/src/main/java/mozilla/components/support/remotesettings/RemoteSettingsClient.kt:66]: accidental commit?

File: fenix/app/src/main/java/org/mozilla/fenix/utils/Settings.kt [hunk 12]
```diff
@@ -767,6 +767,11 @@ class Settings(private val appContext: Context) : PreferencesHolder {
         true,
     )
 
+    val remoteSettingsServer by booleanPreference(
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/utils/Settings.kt:770]: let's update the name of this to reflect it's a boolean
```suggestion
    val useProductionRemoteSettingsServer by booleanPreference(
```

File: focus-android/app/src/main/res/xml/secret_settings.xml [hunk 13]
```diff
@@ -2,10 +2,17 @@
 <!-- This Source Code Form is subject to the terms of the Mozilla Public
    - License, v. 2.0. If a copy of the MPL was not distributed with this
    - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
-<PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android">
+<PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto">
```
Comment (MozillaNoah) [inline focus-android/app/src/main/res/xml/secret_settings.xml:6]: is this XML import still needed?

File: focus-android/app/src/main/java/org/mozilla/focus/settings/advanced/SecretSettingsFragment.kt [hunk 14]
```diff
@@ -26,6 +28,12 @@ class SecretSettingsFragment :
...
         addPreferencesFromResource(R.xml.secret_settings)
+
+        requirePreference<SwitchPreferenceCompat>(R.string.pref_key_remote_server_prod).apply {
+            isVisible = true
+            isChecked = context.settings.remoteSettingsServer
+            onPreferenceChangeListener = SharedPreferenceUpdater()
+        }
     }
 
     override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String?) {
```
Comment (MozillaNoah) [inline focus-android/app/src/main/java/org/mozilla/focus/settings/advanced/SecretSettingsFragment.kt:39]: haven't tested this locally. is this function getting called for the new preference in Focus?

**Review by MozillaNoah — COMMENTED** [review]
the change from the radio button to the switch looks great! just a couple outstanding questions

File: android-components/components/lib/crash/docs/metrics.md (hunk 10 above)
Comment (sarah541) [reply android-components/components/lib/crash/docs/metrics.md:65]: Oops!

File: focus-android/app/src/main/res/xml/secret_settings.xml (hunk 13 above)
Comment (sarah541) [reply focus-android/app/src/main/res/xml/secret_settings.xml:6]: No more, thanks for the catch

File: focus-android/app/src/main/java/org/mozilla/focus/settings/advanced/SecretSettingsFragment.kt (hunk 14 above)
Comment (sarah541) [reply focus-android/app/src/main/java/org/mozilla/focus/settings/advanced/SecretSettingsFragment.kt:39]: It is! Yes. Also gonna add screenshots

**Comment by MozillaNoah** [comment]
Settings code looks good! I'll leave the final review/approval to @MatthewTighe

File: android-components/components/lib/crash/docs/metrics.md [hunk 15]
```diff
@@ -8,57 +8,25 @@ This means you might have to go searching through the dependency tree to get a f
 
 # Pings
 
-- [crash](#crash)
-- [metrics](#metrics)
+- [events](#events)
```
Comment (MatthewTighe) [inline android-components/components/lib/crash/docs/metrics.md:11]: Can we revert the changes to this file?

**Review by MatthewTighe — APPROVED** [review]
LGTM outside of the metrics file
